util/check_dif_statuses.py

```diff
@@ -0,0 +1,314 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Usage:
+#   run "./check_dif_statuses.py repo_top" to generate a status report for the
+#   DIFs. The report is generated by cross referencing the git commit history
+#   of each DIF with that of the HW it actuates to provide OpenTitan developers
+#   with information about what DIFs require updating.
+
+import argparse
+import collections
+import glob
+import io
+import json
+import logging
+import os
+import re
+import sys
+from contextlib import redirect_stdout
+from enum import Enum
+
+import gitfame
+import hjson
+import pydriller
+from tabulate import tabulate
+from termcolor import colored
+
+# The IP suffix of some DIFs are slightly different than the IP directory name
+# in the hw/ip/ directory, so here is the translation for the applicable IPs.
+IP_TRANSLATION = {
+    "plic": "rv_plic",
+    "entropy": "entropy_src",
+}
+TABLE_FORMAT = "grid"
+
+
+class OTComponent(Enum):
+    """Type of OpenTitan component."""
+    DIF = 1
+    HW = 2
+
+
+class DIFStatus:
+    """Hold all DIF status information for displaying."""
+    def __init__(self, repo_top, difs_root_path, dif_name):
+        # Get DIF/IP names and path.
+        self.dif_name = dif_name
+        self.ip = self.dif_name[4:]
+        if self.ip in IP_TRANSLATION:
+            self.ip = IP_TRANSLATION[self.ip]
+        self.dif_path = os.path.join(difs_root_path, dif_name)
```

Comment:
Maybe I'm missing something, but are you suggesting replacing all os.path.join's with pathlib calls? (If so I can definitely make that adjustment.) Or are you saying there is a way for pathlib to resolve partial matches in a potential path? Because the issue here is that for two of the IP blocks (namely, the rv_plic and entropy_src) the IP block names are different from the DIF names (e.g., "dif_plic.*" is the name for the DIF that actuates the "rv_plic" IP block)? If its the latter, that awesome! Could you elaborate more on this? (Admittedly, the way its written now is somewhat of a hack, but the IP block names deviate slightly from the DIF names in some cases, which probably should not be?)

Reply:
> are you suggesting replacing all os.path.join's with pathlib calls?

Not really, unless you really want to. The context that github shows is confusing but my comment was only for `os.path` usage in general. I think `pathlib` is cleaner and just wanted to mention that in case you are not familiar with it. For example,
```
os.path.join(self.hw_path, "data", self.ip + ".prj.hjson")
os.path.isdir(full_path[:-2])
```
would be something like
```
(self.hw_path / "data" / self.ip).with_suffix(".prj.hjson")
full_path.with_suffix('').is_dir()
```

> Or are you saying there is a way for pathlib to resolve partial matches in a potential path?

I'm not sure if it has something that you can use directly here, you could probably use `iterdir()` but wouldn't be cleaner than this.

Reply:
Ah got it. Thanks!